### ISSUE TITLE
Inference information are lost depending on code structure

### ISSUE THREAD

**Issue by bersbersbers** [PyCQA/pylint#4047]
Is your issue fixed on the preview release?: `pip install pylint astroid --pre -U` --> no

### Steps to reproduce
```
"""False-positive no-member."""
import temppathlib


def _func(params=(True,)):
    new_params = temppathlib.gettempdir().glob("*")

    # this changes the type of the elements of params:
    if True and True:
        params = new_params
    else:
        params = new_params

    # with this equivalent code, linting gives no error, as expected:
    # params = new_params if params else new_params

    for parm in params:
        print(parm.name)


_func()
```

### Current behavior
> bug.py:18:14: E1101: Instance of 'bool' has no 'name' member (no-member)

### Expected behavior
> No error.

### pylint --version output
```
pylint 2.6.0
astroid 2.4.2
Python 3.8.6 (default, Sep 29 2020, 15:36:55) 
[GCC 7.5.0]
```

**Comment by Pierre-Sassoulas** [comment]
Closed in favor of #4048

**Comment by bersbersbers** [comment]
#4048 is a different bug in my opinion (false positive vs. false negative).

**Comment by Pierre-Sassoulas** [comment]
You're right, my bad. Here the problem is with astroid not being able to infer types in `temppathlib` . This is similar to #3856 and countless other false positive for `no-member`. 

The following code `b.py`:
```python
"""False-positive no-member."""

class Named:
    def __init__(self, name):
        self.name = name

def _func(params=(True,)):
    new_params = [Named(1), Named(2)]
    if True and True:
        params = new_params
    else:
        params = new_params
    # with this equivalent code, linting gives no error, as expected:
    # params = new_params if params else new_params
    for parm in params:
        print(parm.name)

_func()
```
Properly gives:
`pylint b.py`
``` 
************* Module beu
b.py:9:7: R1727: Boolean condition 'True and True' will always evaluate to 'True' (condition-evals-to-constant)

------------------------------------------------------------------
Your code has been rated at 9.09/10 (previous run: 9.09/10, +0.00)
```

`python3 beu.py`
```
1
2
```

**Comment by bersbersbers** [comment]
Interesting - because `temppathlib.gettempdir()` is properly typed, see
https://github.com/Parquery/temppathlib/commit/88bc2f9d12681c0dd472d96923b739c3aa44762b

Also, `params = new_params if params else new_params` works as expected, so I wonder what the difference is between the four-line and the inline `if`.

**Comment by bersbersbers** [comment]
I think the new title nails it, thank you!

**Comment by bersbersbers** [comment]
Minimal examples:
```
"""False-positive no-member."""
import temppathlib

def _func(params=(True,)):
    new_params = temppathlib.gettempdir().glob("*")

    if params:
        params = new_params
    else:
        params = new_params

    for parm in params:
        print(parm.name)

_func()
```
--> false positive.

```
"""False-positive no-member."""
import temppathlib

def _func(params=(True,)):
    new_params = temppathlib.gettempdir().glob("*")

    params = new_params if params else new_params

    for parm in params:
        print(parm.name)

_func()
```
--> works as expected.
